File: LeetCode/JavaScript/B21-316-Remove Duplicate Letters2.js
/**
 * @param {string} s
 * @return {string}
 */
var removeDuplicateLetters = function (s) {
  const counter = new Map(),
    seen = new Map(),
    stack = new Array();
  Array.from(s).forEach((char) => {
    counter.set(char, counter.get(char) ? counter.get(char) + 1 : 1);
  });

  for (char of s) {
    counter.set(char, counter.get(char) - 1);
    if (seen.get(char)) {
      continue;
    }
    while (!!stack.length && stack[stack.length - 1] > char && !!counter.get(stack[stack.length - 1])) {
      seen.set(stack.pop(), false);
    }

    stack.push(char);
    seen.set(char, true);
  }

  return stack.join("");
};

console.log(removeDuplicateLetters("cbacdcbccbb"));

File: LeetCode/5회차/B21-316-Remove Duplicate Letters2.py
from collections import Counter, defaultdict


class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        counter, seen, stack = Counter(s), defaultdict(bool), []

        for char in s:
            if seen[char]:
                continue
            while stack and stack[-1] > char and counter[stack[-1]] > 1:
                seen[stack.pop()] = False
            stack.append(char)
            seen[char] = True

        return ''.join(stack)
